c2rust-analyze/src/borrowck/mod.rs

```diff
@@ -229,12 +322,68 @@ fn assign_origins<'tcx>(
         } else {
             hypothesis[lty.label]
         };
+
+        let construct_adt_origins = |ty: &Ty, amaps: &mut AtomMaps| -> Option<&_> {
+            let mut fully_derefed_ty = ty;
+            loop {
+                match fully_derefed_ty.kind() {
+                    TyKind::RawPtr(ty) => {
```

Comment:
I don't think this traversal of `ty` is necessary.  `ltcx.relabel` already recurses through all the nodes in the type - for example, if the input is `*mut MyStruct`, the callback will run once for `MyStruct` and again for `*mut MyStruct`, attaching the two labels returned by the callback to their respective nodes.  With the traversal logic here, if `MyStruct` has origin parameters, then I think it will generate two sets of concrete origins, one for the `MyStruct` type itself and another copy attached to the outer `RawPtr`, something like `*mut<'h2, 'h3> MyStruct<'h0, 'h1>`.  But it doesn't make sense to have extra origins at the `*mut` level like this.  I think `assign_origins` should have basically the same structure as on master, but with an additional `TyKind::Adt` case that assigns extra origins only on the struct type itself.

Reply:
addressed in edff552b